[CORE-10237] Implement metrics for shadow link

IoannisRP opened the pull request:
Implements: [CORE-10237](https://redpandadata.atlassian.net/browse/CORE-10237)

| Metric | Type | Description | Labels | Aggregation labels |
| -- | -- | -- | -- | -- |
| `redpanda_shadow_link_total_records_fetched` | count | Total number of records fetched by the replicator | `shadow_link_name`, `shard` |
| `redpanda_shadow_link_total_records_written` | count | Total number of records written by the replicator | `shadow_link_name`, `shard` |
| `redpanda_shadow_link_total_bytes_fetched` | count | Total number of bytes fetched by the replicator | `shadow_link_name`, `shard` |
| `redpanda_shadow_link_total_bytes_written` | count | Total number of bytes written by the replicator | `shadow_link_name`, `shard` |
| `redpanda_shadow_link_shadow_topic_state` | gauge | Number of shadow topics in the respective state | `shadow_link_name` | `shard` |
| `redpanda_shadow_link_client_errors` | count | Number of errors seen by the client |`shadow_link_name`, `shard` |
| `redpanda_shadow_link_shadow_lag` | gauge | Lag of the shadow partition against the source partition |`shadow_link_name`, `topic`, `partition` | `shard` |


</div>


<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.2.x
- [ ] v25.1.x
- [ ] v24.3.x

## Release Notes

* none

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->


[CORE-10237]: https://redpandadata.atlassian.net/browse/CORE-10237?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

IoannisRP committed:
cl/replication: Add probe data type for fetch and replication

IoannisRP committed:
cl/replication: Factor our fetch_data type

This is a refactor-only commit. No functional changes

IoannisRP committed:
cl/replication: add tracking for fetch and replication data in partition replicator

IoannisRP commented on src/v/cluster_link/link_probe.cc:
```diff
@@ -0,0 +1,118 @@
+    const auto sl_name = sl_label(_link.get_config().name);
+
+    // topics per mirror_topic_status
+    auto status_to_metric = std::views::transform(
+      [this, &sl_name](const model::mirror_topic_status status) {
+          auto status_label = sm::label_instance(
+            "status", to_string_view(status));
+          return sm::make_gauge(
+                   "shadow_topic_state",
+                   [this, status] { return count_topics(status); },
+                   sm::description(
+                     "Number of shadow topics in the respective state"),
+                   {sl_name, status_label})
+            .aggregate({sm::shard_label});
+      });
+    constexpr auto statuses = std::to_array({
+      model::mirror_topic_status::active,
+      model::mirror_topic_status::failed,
+      model::mirror_topic_status::paused,
+      model::mirror_topic_status::failing_over,
+      model::mirror_topic_status::failed_over,
+      model::mirror_topic_status::promoting,
+      model::mirror_topic_status::promoted,
+    });
```
i don't think this does what you think it does :sweat_smile:

IoannisRP commented on tests/rptest/tests/cluster_linking_e2e_test.py:
```diff
@@ -1359,3 +1364,177 @@ def test_link_failover(self, with_failures):
+                    "promoted",
+                ]
+                return topics_per_status["active"] == n_active and all(
+                    [topics_per_status[s] == 0 for s in other_statuses]
```
if the samples are not there, it should fail at this stage.

IoannisRP commented:
/dt

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--98f6aad3-5817-4df7-b3e9-83922296b8bd-->
<!--TEST RESULTS-->
<!--c908cbce-1497-4edb-87ff-abf1c59c7789-->
<!--TEST RESULTS-->
<!--004f088a-95a1-43df-b4eb-aff0aa610311-->
<!--TEST RESULTS-->
<!--1bc32ba2-be84-45df-a153-e8fa2439e3e6-->
<!--TEST RESULTS-->
<!--c22a985e-1f3f-48ee-9851-92e005a007e7-->
<!--TEST RESULTS-->
<!--b277f82e-2849-4b78-8cdb-47673f3f7b8e-->
<!--TEST RESULTS-->
<!--b019bb57-b88d-4742-bac3-2fe47e083dcc-->
<!--TEST RESULTS-->
<!--929fb531-f577-4d44-9b75-5cd0944849e2-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/74247">build#74247</a></summary>


| test_class               | test_method                                             | test_arguments                              | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                    | test_history                                                                                                                                                                                   |
|:-------------------------|:--------------------------------------------------------|:--------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:----------------------------------------------------------------------------------------------------------------------------------------------------------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| NodesDecommissioningTest | test_decommissioning_finishes_after_manual_cancellation | {"cloud_topic": true, "delete_topic": true} | integration | https://buildkite.com/redpanda/redpanda/builds/74247#0199e7c1-d7b1-47a1-a1af-05de7f4b8475 | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=NodesDecommissioningTest&test_method=test_decommissioning_finishes_after_manual_cancellation |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/74322">build#74322</a></summary>


| test_class               | test_method                                             | test_arguments                                                    | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 | test_history                                                                                                                                                                                   |
|:-------------------------|:--------------------------------------------------------|:------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| NodesDecommissioningTest | test_decommissioning_finishes_after_manual_cancellation | {"cloud_topic": false, "delete_topic": true}                      | integration | https://buildkite.com/redpanda/redpanda/builds/74322#0199ed5f-2194-4eff-ada5-4ca088cb2695 | FLAKY         | 20/21    | upstream reliability is '97.87234042553192'. current run reliability is '95.23809523809523'. drift is 2.63425 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=NodesDecommissioningTest&test_method=test_decommissioning_finishes_after_manual_cancellation |
| QuickTerminateTest       | test_terminate                                          | null                                                              | integration | https://buildkite.com/redpanda/redpanda/builds/74322#0199ed5f-2190-4817-9d40-1da804d7b8f4 | FLAKY         | 17/21    | upstream reliability is '100.0'. current run reliability is '80.95238095238095'. drift is 19.04762 and the allowed drift is set to 50. The test should PASS            | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=QuickTerminateTest&test_method=test_terminate                                                |
| TimeQueryTest            | test_timequery                                          | {"batch_cache": false, "cloud_storage": true, "spillover": false} | integration | https://buildkite.com/redpanda/redpanda/builds/74322#0199ed5f-1758-4535-882c-682a15777685 | FAIL          | 0/1      |                                                                                                                                                                        | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=TimeQueryTest&test_method=test_timequery                                                     |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/74328">build#74328</a></summary>


| test_class                                             | test_method                                            | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason   | test_history   |
|:-------------------------------------------------------|:-------------------------------------------------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:---------|:---------------|
| MasterTestSuite                                        | unlicensed_reject_defaults                             |                  | unit        | https://buildkite.com/redpanda/redpanda/builds/74328#0199ee23-7720-4ed1-982d-2bc077a472c6 | FAIL          | 0/1      |          |                |
| src/v/cluster/cloud_metadata/tests/offsets_lookup_test | src/v/cluster/cloud_metadata/tests/offsets_lookup_test |                  | unit        | https://buildkite.com/redpanda/redpanda/builds/74328#0199ee23-7720-4ed1-982d-2bc077a472c6 | FAIL          | 0/1      |          |                |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/74516">build#74516</a></summary>


| test_class                    | test_method                                    | test_arguments                                                                    | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  | test_history                                                                                                                                                                          |
|:------------------------------|:-----------------------------------------------|:----------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ShadowLinkBasicTests          | test_rapid_shadow_link_toggling                | null                                                                              | integration | https://buildkite.com/redpanda/redpanda/builds/74516#019a01c5-06e7-4491-b3a3-f11690488194 | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS               | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkBasicTests&test_method=test_rapid_shadow_link_toggling                    |
| ShadowLinkingReplicationTests | test_replication_basic                         | {"shuffle_leadership": true, "source_cluster_spec": {"cluster_type": "redpanda"}} | integration | https://buildkite.com/redpanda/redpanda/builds/74516#019a01c4-d881-4feb-b50f-e3aa0013dc21 | FLAKY         | 16/21    | upstream reliability is '93.15789473684211'. current run reliability is '76.19047619047619'. drift is 16.96742 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingReplicationTests&test_method=test_replication_basic                    |
| ShadowLinkingReplicationTests | test_replication_basic                         | {"shuffle_leadership": true, "source_cluster_spec": {"cluster_type": "redpanda"}} | integration | https://buildkite.com/redpanda/redpanda/builds/74516#019a01c5-06ee-4734-ab0c-69f6263a33c6 | FLAKY         | 18/21    | upstream reliability is '93.15789473684211'. current run reliability is '85.71428571428571'. drift is 7.44361 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingReplicationTests&test_method=test_replication_basic                    |
| ShadowLinkingReplicationTests | test_replication_with_failures                 | null                                                                              | integration | https://buildkite.com/redpanda/redpanda/builds/74516#019a01c4-d882-461a-bf4d-31e531136020 | FLAKY         | 10/21    | upstream reliability is '65.22911051212938'. current run reliability is '47.61904761904761'. drift is 17.61006 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingReplicationTests&test_method=test_replication_with_failures            |
| NodesDecommissioningTest      | test_node_is_not_allowed_to_join_after_restart | {"cloud_topic": true, "new_bootstrap": false}                                     | integration | https://buildkite.com/redpanda/redpanda/builds/74516#019a01c4-d881-4feb-b50f-e3aa0013dc21 | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS               | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=NodesDecommissioningTest&test_method=test_node_is_not_allowed_to_join_after_restart |
| NodesDecommissioningTest      | test_recommissioning_node_finishes             | {"cloud_topic": false}                                                            | integration | https://buildkite.com/redpanda/redpanda/builds/74516#019a01c4-d87c-4a6c-bfb3-7c888b277b4e | FLAKY         | 19/21    | upstream reliability is '98.88888888888889'. current run reliability is '90.47619047619048'. drift is 8.4127 and the allowed drift is set to 50. The test should PASS   | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=NodesDecommissioningTest&test_method=test_recommissioning_node_finishes             |
| TxAtomicProduceConsumeTest    | test_basic_tx_consumer_transform_produce       | {"with_failures": true}                                                           | integration | https://buildkite.com/redpanda/redpanda/builds/74516#019a01c5-06ee-4734-ab0c-69f6263a33c6 | FLAKY         | 20/21    | upstream reliability is '99.1769547325103'. current run reliability is '95.23809523809523'. drift is 3.93886 and the allowed drift is set to 50. The test should PASS   | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=TxAtomicProduceConsumeTest&test_method=test_basic_tx_consumer_transform_produce     |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/74623">build#74623</a></summary>


| test_class                           | test_method                                             | test_arguments                                                               | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                   | test_history                                                                                                                                                                                   |
|:-------------------------------------|:--------------------------------------------------------|:-----------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ShadowLinkConsumeGroupsMirroringTest | test_continuous_group_sync                              | {"source_cluster_spec": {"cluster_type": "redpanda"}, "with_failures": true} | integration | https://buildkite.com/redpanda/redpanda/builds/74623#019a07d7-fc00-42a9-bcc1-143e9cb27b2d | FLAKY         | 19/21    | upstream reliability is '100.0'. current run reliability is '90.47619047619048'. drift is 9.52381 and the allowed drift is set to 50. The test should PASS               | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkConsumeGroupsMirroringTest&test_method=test_continuous_group_sync                  |
| NodesDecommissioningTest             | test_decommissioning_finishes_after_manual_cancellation | {"cloud_topic": true, "delete_topic": true}                                  | integration | https://buildkite.com/redpanda/redpanda/builds/74623#019a07d7-fbf6-4525-a64a-0a65560c2d2d | FLAKY         | 19/21    | upstream reliability is '99.2156862745098'. current run reliability is '90.47619047619048'. drift is 8.7395 and the allowed drift is set to 50. The test should PASS     | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=NodesDecommissioningTest&test_method=test_decommissioning_finishes_after_manual_cancellation |
| ShadowLinkingRandomOpsTest           | test_node_operations                                    | {"failures": false}                                                          | integration | https://buildkite.com/redpanda/redpanda/builds/74623#019a07d7-fbf6-4525-a64a-0a65560c2d2d | FLAKY         | 8/21     | upstream reliability is '94.23076923076923'. current run reliability is '38.095238095238095'. drift is 56.13553 and the allowed drift is set to 50. The test should FAIL | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingRandomOpsTest&test_method=test_node_operations                                  |
| ShadowLinkingRandomOpsTest           | test_node_operations                                    | {"failures": true}                                                           | integration | https://buildkite.com/redpanda/redpanda/builds/74623#019a07d7-fbf7-4de1-86b8-5dcca0ea14c8 | FLAKY         | 3/21     | upstream reliability is '35.84905660377358'. current run reliability is '14.285714285714285'. drift is 21.56334 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingRandomOpsTest&test_method=test_node_operations                                  |
| ShadowLinkingRandomOpsTest           | test_node_operations                                    | {"failures": true}                                                           | integration | https://buildkite.com/redpanda/redpanda/builds/74623#019a07da-a38c-4617-b02c-23e74643bfe4 | FLAKY         | 2/21     | upstream reliability is '100.0'. current run reliability is '9.523809523809524'. drift is 90.47619 and the allowed drift is set to 50. The test should FAIL              | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingRandomOpsTest&test_method=test_node_operations                                  |
| WriteCachingFailureInjectionE2ETest  | test_crash_all                                          | {"use_transactions": false}                                                  | integration | https://buildkite.com/redpanda/redpanda/builds/74623#019a07d7-fc01-4cf5-8653-6595501485e5 | FLAKY         | 16/21    | upstream reliability is '82.07317073170731'. current run reliability is '76.19047619047619'. drift is 5.88269 and the allowed drift is set to 50. The test should PASS   | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=WriteCachingFailureInjectionE2ETest&test_method=test_crash_all                               |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/74706">build#74706</a></summary>


| test_class                       | test_method                | test_arguments        | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                    | test_history                                                                                                                                                              |
|:---------------------------------|:---------------------------|:----------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:--------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:--------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| PartitionMovementTest            | test_static                | {"num_to_upgrade": 0} | integration | https://buildkite.com/redpanda/redpanda/builds/74706#019a0d26-fb06-42e3-bc32-ef36d03afc65 | FLAKY         | 16/21    | upstream reliability is '99.06759906759906'. current run reliability is '76.19047619047619'. drift is 22.87712 and the allowed drift is set to 50. The test should PASS   | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=PartitionMovementTest&test_method=test_static                           |
| ShadowLinkingRandomOpsTest       | test_node_operations       | {"failures": false}   | integration | https://buildkite.com/redpanda/redpanda/builds/74706#019a0d28-cd3b-43bc-add0-8b0ac0f26c88 | FLAKY         | 3/21     | upstream reliability is '37.301587301587304'. current run reliability is '14.285714285714285'. drift is 23.01587 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingRandomOpsTest&test_method=test_node_operations             |
| ShadowLinkingRandomOpsTest       | test_node_operations       | {"failures": true}    | integration | https://buildkite.com/redpanda/redpanda/builds/74706#019a0d26-fb00-4902-a90f-162952752c5f | FLAKY         | 2/21     | upstream reliability is '64.74358974358975'. current run reliability is '9.523809523809524'. drift is 55.21978 and the allowed drift is set to 50. The test should FAIL   | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingRandomOpsTest&test_method=test_node_operations             |
| WriteCachingFailureInjectionTest | test_unavoidable_data_loss | null                  | integration | https://buildkite.com/redpanda/redpanda/builds/74706#019a0d26-fb03-415e-94a1-5981930f0079 | FLAKY         | 19/21    | upstream reliability is '92.2811059907834'. current run reliability is '90.47619047619048'. drift is 1.80492 and the allowed drift is set to 50. The test should PASS     | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=WriteCachingFailureInjectionTest&test_method=test_unavoidable_data_loss |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/74817">build#74817</a></summary>


| test_class            | test_method                                  | test_arguments        | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  | test_history                                                                                                                                                             |
|:----------------------|:---------------------------------------------|:----------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ShadowLinkBasicTests  | test_rapid_shadow_link_toggling              | null                  | integration | https://buildkite.com/redpanda/redpanda/builds/74817#019a1233-75f3-47a7-9495-80dbdf2e8a84 | FLAKY         | 14/21    | upstream reliability is '93.22548028311427'. current run reliability is '66.66666666666666'. drift is 26.55881 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkBasicTests&test_method=test_rapid_shadow_link_toggling       |
| SegmentMsTest         | test_segment_rolling_with_retention_consumer | null                  | integration | https://buildkite.com/redpanda/redpanda/builds/74817#019a1231-5f19-4003-a931-5011367944f2 | FLAKY         | 17/21    | upstream reliability is '92.07003089598352'. current run reliability is '80.95238095238095'. drift is 11.11765 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=SegmentMsTest&test_method=test_segment_rolling_with_retention_consumer |
| PartitionMovementTest | test_empty                                   | {"num_to_upgrade": 0} | integration | https://buildkite.com/redpanda/redpanda/builds/74817#019a1233-75fc-432b-8db0-a58ac070c650 | FLAKY         | 17/21    | upstream reliability is '95.82577132486388'. current run reliability is '80.95238095238095'. drift is 14.87339 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=PartitionMovementTest&test_method=test_empty                           |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/74903">build#74903</a></summary>


| test_class               | test_method                                             | test_arguments                                                                                                 | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  | test_history                                                                                                                                                                                   |
|:-------------------------|:--------------------------------------------------------|:---------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ShadowLinkBasicTests     | test_topic_properties_replicated                        | {"source_cluster_spec": {"cluster_type": "kafka", "kafka_quorum": "COMBINED_KRAFT", "kafka_version": "3.8.0"}} | integration | https://buildkite.com/redpanda/redpanda/builds/74903#019a15f2-fb42-4e14-a0fb-e009b2688d25 | FLAKY         | 19/21    | upstream reliability is '95.578231292517'. current run reliability is '90.47619047619048'. drift is 5.10204 and the allowed drift is set to 50. The test should PASS    | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkBasicTests&test_method=test_topic_properties_replicated                            |
| DataMigrationsApiTest    | test_creating_and_listing_migrations                    | null                                                                                                           | integration | https://buildkite.com/redpanda/redpanda/builds/74903#019a15ee-bbf3-45f0-b7b4-041080b3aba0 | FLAKY         | 14/21    | upstream reliability is '99.84520123839009'. current run reliability is '66.66666666666666'. drift is 33.17853 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DataMigrationsApiTest&test_method=test_creating_and_listing_migrations                       |
| NodesDecommissioningTest | test_decommissioning_finishes_after_manual_cancellation | {"cloud_topic": false, "delete_topic": false}                                                                  | integration | https://buildkite.com/redpanda/redpanda/builds/74903#019a15f2-fb41-40d5-bd7f-6cc76c58539d | FLAKY         | 20/21    | upstream reliability is '99.1044776119403'. current run reliability is '95.23809523809523'. drift is 3.86638 and the allowed drift is set to 50. The test should PASS   | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=NodesDecommissioningTest&test_method=test_decommissioning_finishes_after_manual_cancellation |
| TopicRecoveryTest        | test_many_partitions                                    | {"check_mode": "check_manifest_and_segment_metadata", "cloud_storage_type": 2}                                 | integration | https://buildkite.com/redpanda/redpanda/builds/74903#019a15ee-bbf3-45f0-b7b4-041080b3aba0 | FLAKY         | 20/21    | upstream reliability is '99.77324263038548'. current run reliability is '95.23809523809523'. drift is 4.53515 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=TopicRecoveryTest&test_method=test_many_partitions                                           |
</details>
<!--/body-->

michael-redpanda commented on src/v/cluster_link/link_probe.cc:
```diff
@@ -0,0 +1,118 @@
+int link_probe::count_topics(const model::mirror_topic_status status) const {
+    if (ss::this_shard_id() != ::cluster::controller_stm_shard) {
+        return 0;
+    }
```
so this probe only returns values on shard0 on all brokers.  Is the intention to only have this probe running on the controller leader?

IoannisRP replied:
exactly.
I couldn't find a better way to implement this apart from "let it run everywhere, but deactivate it".

michael-redpanda replied:
> exactly.

which part is exactly? lol

I think the problem here is that it's running on every broker, shard 0, not _just_ the broker/shard that's the controller leader (if that's the intention)

bharathv replied:
this is ok, right? Every broker (shard 0) tracks the metrics, we only look at the data from controller leader or something (in grafana)

IoannisRP replied:
had a discussion offline with @michael-redpanda and decided to keep these metrics only on the controller leader.

IoannisRP committed:
k/c/direct_consumer: Add direct consumer probe

IoannisRP committed:
k/c/direct_consumer: Add optional probe to the direct consumer

IoannisRP committed:
cluster_link: Create direct consumer probe in the cl service

bharathv commented on src/v/cluster_link/replication/types.h:
```diff
@@ -15,11 +15,30 @@
+    }
+};
+
+struct probe_data {
```
naming nit:  shoudln't this read "data_probe" instead of "probe_data", the latter sounds like a verb

bharathv commented on src/v/cluster_link/replication/partition_replicator.cc:
```diff
@@ -177,6 +185,10 @@ ss::future<> partition_replicator::fetch_and_replicate() {
                 continue;
             }
+            if (_probe_data) {
+                _probe_data->total_fetched += fetch_counters::from_fetch_data(
```
nit: do from_fetch_data{..) only once? Since it loops through batches for record count, pass the output to replicate_and_wait()

bharathv commented on src/v/cluster_link/link.cc:
```diff
@@ -89,13 +90,20 @@ link::link(
+    _probe = std::make_unique<link_probe>(*this);
+
     co_await _cluster_connection->start();
+    _replication_mgr.set_probe_data(&_probe->replication_data());
```
can we pass the probe to start() or something?

bharathv commented on src/v/cluster_link/link.cc:
```diff
@@ -114,6 +122,7 @@ ss::future<> link::stop() noexcept {
     _task_reconciler.cancel();
     _as.request_abort();
     co_await _replication_mgr.stop();
+    _replication_mgr.reset_probe_data();
```
can we do this in replication_mgr.stop()?

bharathv commented on src/v/cluster_link/link_probe.cc:
```diff
@@ -0,0 +1,118 @@
+        "total_bytes_written",
+        [this] { return _replication_data.total_written.n_bytes; },
+        sm::description(
+          "The total number of bytes written by the sharded replicator"),
```
nit: consider removing "sharded" from the desc (it is an implementation detail and this is public facing)

IoannisRP committed:
k/c/direct_consumer: Add error tracking on fetcher

IoannisRP commented:
Changes in [force-push](https://github.com/redpanda-data/redpanda/compare/b3239c236db63ab12e1688915218e7ca3162a584..4895ecd9a3b43195e74be81b38a54ca9cbe65305):
- add direct consumer probe that tracks errors

IoannisRP commented:
Changes in [force-push](https://github.com/redpanda-data/redpanda/compare/4895ecd9a3b43195e74be81b38a54ca9cbe65305..a8e0ccd526a12d868b5db288735d615ee7353396):
- Update metrics description
- addressed pr comments

IoannisRP commented:
/dt

bharathv reviewed (dismissed):
lgtm.

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/74322-->
<!--95335a3a-e230-460c-ba51-b6884d7727ba-->
<!--title-->
## Retry command for [Build#74322](https://buildkite.com/redpanda/redpanda/builds/74322)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/timequery_test.py::TimeQueryTest.test_timequery@{"batch_cache":false,"cloud_storage":true,"spillover":false}
```
<!--/body-->

IoannisRP commented:
changes in [force-push](https://github.com/redpanda-data/redpanda/compare/a8e0ccd526a12d868b5db288735d615ee7353396..a138980a4bdb72d1f6fd402b36fac4ed94ee89f6):
- change statuses to only be reported by the controller leader

IoannisRP commented:
changes in [force-push](https://github.com/redpanda-data/redpanda/compare/a138980a4bdb72d1f6fd402b36fac4ed94ee89f6..d77abebc800df506301be593db3fd7b7c321dca0):
- make status metrics available only on the controller leader.

IoannisRP commented:
/dt

IoannisRP commented:
changes in [force-push](https://github.com/redpanda-data/redpanda/compare/d77abebc800df506301be593db3fd7b7c321dca0..0370fd8746bcd0bc8fc7df581b87664373c05b9f):
- rebased to dev

IoannisRP committed:
cl/replication: Add a probe to be added to each partition replicator

IoannisRP committed:
cl/replication: Add an optional probe to each partition replicator

IoannisRP committed:
cl/replication: Wire partition replicator probe and data through manager

IoannisRP committed:
cluster_link: Add cluster link probe type

IoannisRP committed:
cluster_link: Add probes to shadow link

This enables:
- a link_probe that exists on the link itself
- a probe for every partition replicator

IoannisRP commented:
Changes in [force-push](https://github.com/redpanda-data/redpanda/compare/0370fd8746bcd0bc8fc7df581b87664373c05b9f..11bb84f27d00ae8fe56d463a4ff96fc89f91628f):
- reworked the `data_probe` section. As the shadow lag needs to be per topic/partition, it makes sense to live on a probe directly on the partition replicator. Since there is a probe, it doesn't make sense to use injection for the rest of the metrics. All the metrics that were using the injected `data_probe` now live in the partition replicator probe, as well.

IoannisRP committed:
cl/replication: wire direct consumer probe configuration through mux consumer

IoannisRP commented:
changes in [force-push](https://github.com/redpanda-data/redpanda/compare/3482aa4cd2a45b43c2f2d99116ffb6375107657b..c03b2d8a3418686da444b0656e3ac8f033adaa42):
- rebase to `dev`

bharathv commented on src/v/cluster_link/replication/partition_replicator.cc:
```diff
@@ -201,6 +201,21 @@ partition_replicator::get_partition_offsets_report() const {
     };
 }
 
+kafka::offset partition_replicator::get_partition_lag() const {
```
should this be an offset_delta?

IoannisRP replied:
good point. Done

bharathv replied:
lost the comment about offset_delta, didn't realize it was only meant for offset_translation, still feels like an artificial limitation tho. Either way fine..I't be nicer to represent lag as int64 or have a dedicated type for it. Using a kafka::offset still seems a bit odd to me (not suggesting to do in this PR, a follow up is fine).

IoannisRP replied:
> or have a dedicated type for it.

Yes. The `offset_delta` :sweat_smile: 

i saw a few different places where deltas are represented by an `offset` type. That's why I went with it to begin with.

I agree with you that using the `offset_delta` for something different than, well... deltas, feels off.

Happy to revisit, if needed.

bharathv reviewed (commented):
lgtm one minor comment.

IoannisRP commented:
changes in [force-push](https://github.com/redpanda-data/redpanda/compare/c03b2d8a3418686da444b0656e3ac8f033adaa42..2b0dc6a9d4e53931e30fa95c345ae1d70ac7d27a):
- change lag type to `kafka::offset_delta`

IoannisRP commented:
changes in [force-push](https://github.com/redpanda-data/redpanda/compare/2b0dc6a9d4e53931e30fa95c345ae1d70ac7d27a..417f62744c960ba8b4ac02f6ee96f91f7f789a6a):
- rebase to `dev` to fix merge conflict... again.

bharathv commented on src/v/cluster_link/replication/replication_probe.cc:
```diff
@@ -0,0 +1,74 @@
+    std::vector<sm::label_instance> total_counters_labels = {
+      topic_label(ntp.tp.topic()),
+      partition_label(ntp.tp.partition()),
+    };
```
sorry missed in the first pass, do we need a link_name/link_id label?

IoannisRP replied:
we add the `shadow_link_name` in the next line 👍

bharathv commented on tests/rptest/tests/cluster_linking_e2e_test.py:
```diff
@@ -1888,6 +1888,9 @@ def check_bytes_written(samples: list[dict[str, MetricSamples]]):
 
+        def check_shadow_lag(node_samples: list[dict[str, MetricSamples]]) -> bool:
+            return check_metric_exists(node_samples, "shadow_lag")
+
```
probably useful to add more tests to ensure lag is growing and shrinking as needed (across leadership changes etc). Its ok to add as a follow on change but I think such a test is really helpful.

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/74623-->
<!--d816cafd-66d1-4419-982f-ce301b334c99-->
<!--f1cc8be7-6429-4693-9f00-f9b829046ebf-->
<!--title-->
## Retry command for [Build#74623](https://buildkite.com/redpanda/redpanda/builds/74623)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/shadow_linking_rnot_test.py::ShadowLinkingRandomOpsTest.test_node_operations@{"failures":true}
tests/rptest/tests/shadow_linking_rnot_test.py::ShadowLinkingRandomOpsTest.test_node_operations@{"failures":false}
```
<!--/body-->

IoannisRP commented:
changes in [force-push](https://github.com/redpanda-data/redpanda/compare/417f62744c960ba8b4ac02f6ee96f91f7f789a6a..5f85d8394475faeac26c06f89a3130ababe38108):
- rebased to `dev`

IoannisRP commented:
Changes in [force-push](https://github.com/redpanda-data/redpanda/compare/5f85d8394475faeac26c06f89a3130ababe38108..c827f16e7efca2fd8d36ed731a0336e78be8ac53):
- added test for lag

michael-redpanda commented on src/v/cluster_link/replication/types.cc:
```diff
@@ -0,0 +1,29 @@
+    ret.n_records = std::ranges::fold_left(
+      data.batches | std::views::transform(&model::record_batch::record_count),
+      int32_t{},
+      std::plus<int32_t>{});
```
Yeah record_count is an `int32_t`...

joe-redpanda replied:
record count is an int32, so the sum of multiple record counts should be int64

IoannisRP replied:
@joe-redpanda is it probable that the records of a single record_batch could overflow?
Note that the result is being stored into a size_t.

joe-redpanda replied:
The intermediate sizes have to be large enough as well:
https://godbolt.org/z/86szch5nK

IoannisRP replied:
i don't disagree that any integer addition can overflow. I was asking whether overflowing is a concern in this specific use case.

Anyways... updated to `int64_t`

michael-redpanda commented on src/v/cluster_link/replication/replication_probe.cc:
```diff
@@ -0,0 +1,74 @@
+          [this] { return _total_written.n_bytes; },
+          sm::description("Total number of bytes written by the replicator"),
+          total_counters_labels)
+          .aggregate(total_counters_agg),
```
So `counter` here initially makes sense, but I'm wondering if this needs to be a gauge.  As leadership moves, this metric will move and _reset_ back to zero.  I'm wondering because of this behavior it needs to be a gauge because metric consumers would become confused as to why the value went backwards.  cc'ing @BenPope to keep me honest here (happy to be wrong! :) )

IoannisRP replied:
thinking about it, it would also go down if we pause a topic in the link, as it will stop its replicators.
Are we ok with making a gauge? The alternative is to put the data counters on the link, as it was before :sweat:

IoannisRP replied:
Had an offline discussion. Moving the data to the `link` level, so these can properly be counters.

michael-redpanda commented on src/v/cluster_link/link_probe.h:
```diff
@@ -0,0 +1,44 @@
+/*
+ * Copyright 2025 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
```
chore: this should be RCL licensed

michael-redpanda commented on src/v/kafka/client/direct_consumer/direct_consumer_probe.cc:
```diff
@@ -0,0 +1,52 @@
+/*
+ * Copyright 2025 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
```
note: this is fine being bsl

michael-redpanda reviewed (commented):
looking nice!

IoannisRP committed:
cl/replication: expose partition shadow lag function in replicator

IoannisRP commented:
changes in [force-push](https://github.com/redpanda-data/redpanda/compare/be6f5b9fedd91b4f89a06cef681946c7ff2785c1..5d75398f0257c9c98b289a705c9b46770af6147b):
- moved back to the data injection approach, to ensure that the total counters are monotonically increasing.

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/74706-->
<!--69d46759-9861-4063-ac60-7cf7e98f30dd-->
<!--title-->
## Retry command for [Build#74706](https://buildkite.com/redpanda/redpanda/builds/74706)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/shadow_linking_rnot_test.py::ShadowLinkingRandomOpsTest.test_node_operations@{"failures":true}
```
<!--/body-->

IoannisRP commented on src/v/cluster_link/replication/partition_replicator.h:
```diff
@@ -71,13 +74,17 @@ class partition_replicator {
+    void set_data_probe(link_data_probe*);
+    void unset_data_probe();
+
+    kafka::offset_delta get_partition_lag() const;
```
@bharathv 

https://github.com/redpanda-data/redpanda/blob/7b26db22c9e551ee15752bb6b7c320231ee5f5c0/src/v/model/fundamental.h#L218-L220

I am a bit confused... are you sure that `kafka::offset_delta` the correct type to be used here?

mmaslankaprv replied:
yeah, looks like this should be simply `kafka::offset`

mmaslankaprv replied:
Would be great to cover this with unit tests

IoannisRP replied:
added basic unit test.

mmaslankaprv commented on src/v/cluster_link/replication/types.cc:
```diff
@@ -22,4 +26,14 @@ fmt::iterator partition_offsets_report::format_to(fmt::iterator it) const {
+    ret.n_bytes = data.units.count();
+    ret.n_records = std::ranges::fold_left(
+      data.batches | std::views::transform(&model::record_batch::record_count),
+      int32_t{},
```
nit: initialize to 0 explicitly

IoannisRP replied:
I am not sure what you are referring to. The algorithm is initialized with `0` by `int32_t{}`

IoannisRP commented:
Changes in [force-push](https://github.com/redpanda-data/redpanda/compare/5d75398f0257c9c98b289a705c9b46770af6147b..fbd9777f842cd696df9f6695c1dbf89070fb7727):
- rebased to `dev` to fix merge conflic

Changes in [force-push](https://github.com/redpanda-data/redpanda/compare/fbd9777f842cd696df9f6695c1dbf89070fb7727..b26e93b11fbefbb8b2945484d04f0f9bdac71c63):
- changed `offset_delta` to `offset`

mmaslankaprv commented on src/v/cluster_link/link_probe.cc:
```diff
@@ -0,0 +1,147 @@
+}
+
+void link_probe::handle_on_leadership_change(
+  ::model::ntp ntp, ntp_leader is_ntp_leader) {
```
const ref ?

mmaslankaprv commented on src/v/cluster_link/link_probe.cc:
```diff
@@ -0,0 +1,147 @@
+          sm::description("Total number of records written by the replicator"),
+          {sl_name})
+          .aggregate({sm::shard_label}),
+        sm::make_counter(
```
there is a special metric type: `make_total_bytes`

mmaslankaprv commented on src/v/cluster_link/replication/partition_replicator.h:
```diff
@@ -101,6 +108,8 @@ class partition_replicator {
       max_in_flight_requests, "partition_replicator"};
     backoff_policy _backoff_policy;
+    std::optional<replication_probe> _probe;
+    link_data_probe* _link_data_probe;
```
can this be replaced by shared ptr ?

mmaslankaprv commented on src/v/cluster_link/replication/replication_probe.cc:
```diff
@@ -0,0 +1,58 @@
+          "shadow_lag",
+          [this] { return _partition_replicator.get_partition_lag(); },
+          sm::description(
+            "Lag of the shadow partition against the source partition"),
```
not necessary copy

IoannisRP replied:
i am not sure what copy you are referring to

mmaslankaprv commented on src/v/cluster_link/replication/replication_probe.cc:
```diff
@@ -0,0 +1,58 @@
+  configuration config, ::model::ntp ntp, partition_replicator& pr)
+  : _partition_replicator{pr}
+  , _public_metrics{} {
+    setup_metrics(std::move(config), ntp);
```
can we use shared ptr here ?

IoannisRP replied:
this is a pointer to the parent object.
The partition replicator creates a replication probe and forwards `*this`

IoannisRP commented:
changes in [force-push](https://github.com/redpanda-data/redpanda/compare/b26e93b11fbefbb8b2945484d04f0f9bdac71c63..ffad662be68e3c5dc70310bf561877b9a7e12fbc):
- addressed pr comments

IoannisRP commented:
Changes in [force-push](https://github.com/redpanda-data/redpanda/compare/ffad662be68e3c5dc70310bf561877b9a7e12fbc..9a900ec68c09da3903c950035edfc476913c6cf9):
- change summing algorithm to `int64_t`

changes in [force-push](https://github.com/redpanda-data/redpanda/compare/9a900ec68c09da3903c950035edfc476913c6cf9..e81768f20c0b408b7cd4bf9b5c4b6e58547af662):
- rebase to `dev` to fix merge conflicts

changes in [force-push](https://github.com/redpanda-data/redpanda/compare/e81768f20c0b408b7cd4bf9b5c4b6e58547af662..22a89d816708bf3b981b88fd4c3a62e9a1f9f6d5):
- fix failing shadow lag unit test

StephanDollberg commented on src/v/cluster_link/replication/replication_probe.cc:
```diff
@@ -0,0 +1,58 @@
+    };
+    std::ranges::copy(config.labels, std::back_inserter(labels));
+
+    std::vector<sm::label> total_counters_agg = {topic_label, partition_label};
```
This is unused?

StephanDollberg reviewed (commented):
> shadow_link_name

Is this just one per target cluster or one per partition?

Adding the per topic/partition metrics on the public interface is problematic as no aggregation is happening there. 

Are these metrics required from a user POV?

IoannisRP commented:
@StephanDollberg 

> Is this just one per target cluster or one per partition?

`shadow_link_name` is one per shadow link. It doesn't change per partition.
For now, a cluster can only have one link.

> Are these metrics required from a user POV?

For the lag, I am gonna defer this to @michael-redpanda | @tyson-redpanda

Would the situation be better if we had an aggregated view (or something like the `max` + `sum` that we have in consumer groups) in the public metrics and a non aggregated view in the internal endpoint?

StephanDollberg commented:
> Would the situation be better if we had an aggregated view (or something like the max + sum that we have in consumer groups) in the public metrics and a non aggregated view in the internal endpoint?

That would be better yes (though it depends on user expectations). If there is no hard requirement the best solution would be to have an aggregated metric (not using metric aggregation internally) and then provide the per partition/topic view via admin endpoints.
